src/modelfree/shaping_wrappers.py

```diff
@@ -0,0 +1,208 @@
+from collections import defaultdict
+import json
+
+import numpy as np
+from stable_baselines import logger
+from stable_baselines.common.base_class import BaseRLModel
+from stable_baselines.common.vec_env import VecEnvWrapper
+
+from modelfree import envs  # noqa - needed to register sumo_auto_contact
+from modelfree.scheduling import LinearAnnealer
+
+
+class RewardShapingEnv(VecEnvWrapper):
+    """A more direct interface for shaping the reward of the attacking agent."""
+
+    default_shaping_params = {
+        'reward_center': 1,
+        'reward_ctrl': -1,
+        'reward_contact': -1,
+        'reward_survive': 1,
+
+        # sparse reward field as per gym_compete/new_envs/multi_agent_env:151
+        'reward_remaining': 1,
+        # dense reward field as per gym_compete/new_envs/agents/humanoid_fighter:45
+        # 'reward_move': 10,
+    }
+
+    def __init__(self, env, shaping_params=default_shaping_params,
+                 reward_annealer=None):
+        super().__init__(env)
+        self.shaping_params = shaping_params
+        self.reward_annealer = reward_annealer
+        self.counter = 0
+
+        self.ep_rew_dict = defaultdict(list)
+        self.step_rew_dict = defaultdict(lambda: [[] for _ in range(self.num_envs)])
+
+    def _log_sparse_dense_rewards(self):
+        if self.counter == 2048 * self.num_envs:
+            num_episodes = len(self.ep_rew_dict['reward_remaining'])
+            dense_terms = ['reward_center', 'reward_ctrl', 'reward_contact', 'reward_survive']
+            for term in dense_terms:
+                assert len(self.ep_rew_dict[term]) == num_episodes
+
+            ep_dense_mean = sum([sum(self.ep_rew_dict[t]) for t in dense_terms]) / num_episodes
+            logger.logkv('epdensemean', ep_dense_mean)
+
+            ep_sparse_mean = sum(self.ep_rew_dict['reward_remaining']) / num_episodes
+            logger.logkv('epsparsemean', ep_sparse_mean)
+
+            c = self.reward_annealer()
+            ep_rew_mean = c * ep_dense_mean + (1 - c) * ep_sparse_mean
+            logger.logkv('eprewmean_true', ep_rew_mean)
+
+            logger.logkv('rew_anneal', c)
+
+            for rew_type in self.ep_rew_dict:
+                self.ep_rew_dict[rew_type] = []
+            self.counter = 0
+
+    def reset(self):
+        return self.venv.reset()
+
+    def step_wait(self):
+        obs, rew, done, infos = self.venv.step_wait()
+
+        # replace rew with differently shaped rew
+        # victim is agent 0, attacker is agent 1
```

Comment:
Can you make the attacker agent index a parameter to `__init__`? Right now all our experiments have it be index `1`, but this could change in the future (e.g. maybe we want to attack the goalie in `KickAndDefend-v0` rather than the shooter).